webapp/channels/src/workers/sw.js

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// Update this every time code is changed.
+// This is needed to delete old cache entries.
+const cacheName = 'v1';
+const HTML_REQUEST = new Request('root.html');
+
+const putInCache = async (request, response) => {
+    const cache = await caches.open(cacheName);
+    if (request.destination === 'document') {
+        await cache.put(HTML_REQUEST, response);
+    } else {
+        await cache.put(request, response);
+    }
+};
+
+const handleCacheRequest = async (request) => {
+    // Using the stale-while-revalidate approach for cached requests to keep them in sync.
+    let responseFromCache;
+    if (request.destination === 'document') {
+        // If it's an HTML document, then we check for a separate key.
+        responseFromCache = await caches.match(HTML_REQUEST);
+    } else {
+        responseFromCache = await caches.match(request);
+    }
+    if (responseFromCache) {
+        // Non-blocking cache insertion.
+        fetchFromNetwork(request);
+        return responseFromCache;
+    }
+
+    // Wait till we get a response.
+    const responseFromNetwork = await fetchFromNetwork(request);
+    return responseFromNetwork;
+};
+
+const fetchFromNetwork = async (request) => {
+    try {
+        const responseFromNetwork = await fetch(request.clone());
+        putInCache(request, responseFromNetwork.clone());
+        return responseFromNetwork;
+    } catch (error) {
+        // eslint-disable-next-line no-console
+        console.error(`error while fetching from sw: ${error}`);
+
+        // return a Response object if network request fails.
+        return new Response('Network error happened', {
+            status: 408,
+            headers: {'Content-Type': 'text/plain'},
+        });
```

Comment:
Asking because I don't know: Is there any potential detail in the network error that we can return?

Reply:
Thanks!